### DISCUSSION TITLE
lesson11 when test svgToImageURI， shows assertion failed.

### DISCUSSION THREAD

**Discussion by RaferGuo** [Cyfrin/foundry-full-course-cu#1507] [Q&A] [answered]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/be4c43fd-75c8-453d-96de-5c5988c2928c)
here is my deploymoodnft.s.sol
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script} from "lib/forge-std/src/Script.sol";
import {MoodNft} from "../src/MoodNft.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";

contract DeployMoodNft is Script {
     
     function run() external returns(MoodNft) {
     }

     function svgToImageURI(string memory svg) public pure returns(string memory) {
        //example
        //input:<svg width="1024px" height = "1024px" ....>
        //outpute:data:image/svg+xml;base64,PD94bWwg...
        string memory baseURL ="data:image/svg+xml;base64,";
        string memory svgBase64Encode = Base64.encode(bytes(string(abi.encodePacked(svg))));
        return string(abi.encodePacked(baseURL,svgBase64Encode));
     }
 }
```
here is mydeployMoodNftTest.t.sol
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Test,console} from "forge-std/Test.sol";
import {DeployMoodNft} from "script/DeployMoodNft.s.sol";

contract DeployMoodNftTest is Test {
    DeployMoodNft public deployer;

    function setUp() public {
        deployer = new DeployMoodNft();

    }

    function testConvertSvgToUri() public {
        string memory expectedUri = "data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1MDAiIGhlaWdodD0iNTAwIj4KPHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5oaSF5b3UgZGVjb2RlZCB0aGlzISA8L3RleHQ+Cjwvc3ZnPg==";
        string memory svg = '<svg xmlns="http://www.w3.org/2000/svg" width="500" height="500"><text x="0" y="15" fill="black">hi!you decoded this! </text></svg>';
        string memory actualUri = deployer.svgToImageURI(svg);
        assert(
            keccak256(abi.encodePacked(actualUri)) == 
            keccak256(abi.encodePacked(expectedUri))
        );
    }
}
```
moodnft.sol
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";

contract MoodNft is ERC721{
    uint256 public s_tokenCounter;
    string private s_sadSvgImageUri;
    string private s_happtSvgImageUri;
    
    enum Mood{
        HAPPY,
        SAD
    }

    mapping(uint256 => Mood) s_tokenIdToMood;
      
    //here is imageUri not tokenUri
    constructor(string memory sadSvgImageUri, string memory happySvgImageUri) ERC721("Mood NFT", "MM"){
        s_tokenCounter = 0;
        s_sadSvgImageUri = sadSvgImageUri;
        s_happtSvgImageUri = happySvgImageUri;
    } 

    function mintNft() public {
        _safeMint(msg.sender, s_tokenCounter);
        s_tokenIdToMood[s_tokenCounter] = Mood.HAPPY;
        s_tokenCounter++;
    }   
    
    //to json
    function _baseURI() internal pure override returns (string memory) {
        return "data:application/json;base64,";
    }

    //tokenURI -> metadata
    function tokenURI(uint256 tokenId) public view override returns(string memory) {
        string memory imageURI;
        if(s_tokenIdToMood[tokenId] == Mood.HAPPY){
            imageURI = s_happtSvgImageUri;
        } else {
            imageURI = s_sadSvgImageUri;
        }
        
        return 
          string(
            abi.encodePacked(//hash
            _baseURI(),
              Base64.encode(//base64
                bytes(//bytes
                  //string
                  abi.encodePacked('{"name": "',name(),'", "description": "An NFT that reflects the owners mood.", "attributes": [{"trait_type": "moodiness", "value": "100"}], "image":"',imageURI,'"}'))
                )
              )
            );
    }
}
```
example's base64
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/a53b008e-46a3-4130-8c6b-d42d8635994a)

**Comment by EngrPips** [comment]
Hello @RaferGuo , can you please push this codebase to github and share the link with me. so i can set it up locally and try debug it. And i am actually wondering why didn't you read the svg string into your test contract directly from your svg file as done in the tutorial. In the process of copying and pasting the svg string you can add an additional space character or remove it and even the slightest change would be accounted for in a base64 encoding which would basically make the base64 encoding output significantly different from each other.

**Reply by RaferGuo** [reply]
here is my git repo：
https://github.com/RaferGuo/foundry-nft-test 
and i found when i paste imageUri to string，i had to delate some space，and make them in one line，or it shows below.maybe  this is mistake，but i have no way to paste it in one line directly，no matter use alt+z or not;
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/29f9affc-d25b-423d-bb68-8131eeae74e4)
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/771ffc27-365b-4338-9cb2-6527d14e5b6b)

**Reply by EngrPips** [reply]
Thanks for sharing the repo, i will set it up try debug it and get back to you with whatever i find.

**Reply by EngrPips** [reply, marked as answer]
Hello @RaferGuo , Below is the result of me running your test on my local machine.

<img width="1440" alt="Screenshot 2024-03-22 at 3 31 36 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/810de13c-001f-49cd-8af0-b2b54fe53958">

So like i said earlier base4 encoding is extremely sensitive and any slightest change in the data you are encoding would be accounted for. what i would advice open your `example.svg` and edit the file so all of the svg code is on a single line as shown below 

<img width="1090" alt="Screenshot 2024-03-22 at 3 37 01 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/57642f3d-bc4a-48b9-8f03-fd417c62b84a">

Then use the base64 encoding command on the file data to get a base64 string output , attach that output string to your `expectedUri` as shown below 

```javascript
string memory expectedUri = "data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1MDAiIGhlaWdodD0iNTAwIj48dGV4dCB4PSIwIiB5PSIxNSIgZmlsbD0iYmxhY2siPmhpIXlvdSBkZWNvZGVkIHRoaXMhIDwvdGV4dD48L3N2Zz4=";
```
Then run your test again, your test should pass. Let me know how that goes and i am looking forward to hear back from you.

**Reply by RaferGuo** [reply]
edit the file to make the svg code  on a single line，then it works. thx
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/5f218402-0777-451a-90ec-ebf0ccd97281)
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/31519376/bae2d2e3-af83-4637-b73f-8d97a23d9141)
should i do the same to put happy.svg or sad.svg in one line in later course， if i meet some similar trouble？

**Reply by EngrPips** [reply]
Yeah you should do the same but if the problem persist, i am always on here.
